handler/src/main/java/io/netty/handler/ssl/OptionalSslHandler.java

```diff
@@ -0,0 +1,111 @@
+import java.util.List;
+
+/**
+ * {@link OptionalSslHandler} is a utility decoder to support both SSL and non-SSL handlers
+ * based on the initial message.
+ */
+public class OptionalSslHandler extends ByteToMessageDecoder {
+
+    private final SslContext sslContext;
+
+    public OptionalSslHandler(SslContext sslContext) {
+        this.sslContext = ObjectUtil.checkNotNull(sslContext, "sslContext");
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext context, ByteBuf in, List<Object> out) throws Exception {
+        if (in.readableBytes() < SslUtils.SSL_RECORD_HEADER_LENGTH) {
+            return;
+        }
+        if (SslHandler.isEncrypted(in)) {
+            handleSsl(context);
+        } else {
+            handleNonSsl(context);
+        }
+    }
+
+    private void handleSsl(ChannelHandlerContext context) {
+        SslHandler sslHandler = null;
+        try {
+            sslHandler = newSslHandler(context, sslContext);
+            context.pipeline().replace(this, newSslHandlerName(), sslHandler);
+            sslHandler = null;
+        } finally {
+            // Since the SslHandler was not inserted into the pipeline the ownership of the SSLEngine was not
+            // transferred to the SslHandler.
+            if (sslHandler != null) {
+                ReferenceCountUtil.safeRelease(sslHandler.engine());
+            }
+        }
+    }
+
+    private void handleNonSsl(ChannelHandlerContext context) {
+        ChannelHandler handler = newNonSslHandler(context);
+        if (handler != null) {
+            context.pipeline().replace(this, newNonSslHandlerName(), handler);
+        } else {
+            context.pipeline().remove(this);
+        }
+    }
+
+    /**
+     * Optionally specify the SSL handler name, this method may return {@code null}.
+     * @return the name of the SSL handler.
+     */
+    protected String newSslHandlerName() {
+        return null;
+    }
+
+    /**
+     * Override to configure the SslHandler eg. enable HttpsEndpointIdentificationAlgorithm.
+     * @param context the {@link ChannelHandlerContext} to use.
+     * @param sslContext the {@link SSLContext} to use.
+     * @return the {@link SslHandler} to be added to the pipeline if the traffic is SSL.
+     */
+    protected SslHandler newSslHandler(ChannelHandlerContext context, SslContext sslContext) {
+        return sslContext.newHandler(context.alloc());
+    }
+
+    /**
+     * Optionally specify the non-SSL handler name, this method may return {@code null}.
+     * @return the name of the non-SSL handler.
+     */
+    protected String newNonSslHandlerName() {
+        return null;
+    }
+
+    /**
+     * Override to configure the ChannelHandler.
+     * @param context the {@link ChannelHandlerContext} to use.
+     * @return the {@link ChannelHandler} to with which the {@link OptionalSslHandler} will be replaced or {@code null}
+     * to just remove the {@link OptionalSslHandler}.
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "context". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)